src/landscape.cpp

```diff
@@ -1321,75 +1323,72 @@ static void BuildRiver(TileIndex begin, TileIndex end, TileIndex spring, bool ma
  */
 static std::tuple<bool, bool> FlowRiver(TileIndex spring, TileIndex begin, uint min_river_length)
 {
-	uint height_begin = TileHeight(begin);
-
 	if (IsWaterTile(begin)) {
 		return { DistanceManhattan(spring, begin) > min_river_length, GetTileZ(begin) == 0 };
 	}
 
-	std::set<TileIndex> marks;
-	marks.insert(begin);
+	int height_begin = TileHeight(begin);
 
-	/* Breadth first search for the closest tile we can flow down to. */
-	std::list<TileIndex> queue;
+	static std::unordered_set<TileIndex::BaseType> marks;
```

Comment:
Why? If you need a hash, maybe implement `std::hash` for `StrongType` as that will be much cleaner in the long run.

Especially the casting of `TileIndex` to `Tile` to `uint` below is nasty, as you actually mean e.g. `begin.base()`. But again, just implement `std::hash` for strong types. It'll be something along the line of (totally untested):
```
template <typename T, std::enable_if_t<std::is_base_of<StrongTypedefBase, T>::value, int> = 0>
struct std::hash<T>
{
    std::size_t operator()(const T& t) const noexcept
    {
        return std::hash(t.base());
    }
};
```

Reply:
I added a commit with the hash implementation before the main commit.